### DISCUSSION TITLE
Motion Planning Support

### DISCUSSION THREAD

**Discussion by uwe-koeckemann** [aiplan4eu/unified-planning#284] [Ideas]
Extending problem and actions with information required for basic combined task and motion planning.

The first part of the example below is pure UP and a simple extension of one of the UPs test cases: a robot and two locations with a single move operator.

```python
import unified_planning as up
from unified_planning.model import Fluent, InstantaneousAction, Object, Problem
from unified_planning.shortcuts import UserType, BoolType, ObjectExp

Location = UserType("location")
Robot = UserType("robot")
robot_at = Fluent("robot_at", BoolType(), robot=Robot, position=Location)

move = InstantaneousAction("move", robot=Robot, l_from=Location, l_to=Location)
robot = move.parameter("robot")
l_from = move.parameter("l_from")
l_to = move.parameter("l_to")
move.add_precondition(robot_at(robot, l_from))
move.add_effect(robot_at(robot, l_from), False)
move.add_effect(robot_at(robot, l_to), True)

l1 = Object("l1", Location)
l2 = Object("l2", Location)
r1 = Object("r1", Location)

problem = Problem("robot")
problem.add_fluent(robot_at)
problem.add_action(move)
problem.add_object(l1)
problem.add_object(l2)
problem.set_initial_value(robot_at(l1), True)
problem.set_initial_value(robot_at(l2), False)
problem.add_goal(robot_at(l2))
```

For the extension, we assume fixed frames in 2D. We attach poses and frames to objects and motion requirements to operators. The robots motion model is a ReedsShepp car which can move forwards and backwards (and stand still) and turn with a minimum turning radius. 

```python
class ReedsSheppCar:
    def __init__(self, turning_radius=4.0):
        self.turning_radius = turning_radius
```

We attach a map to the problem and create a frame and attach it to the robot. We also attach a motion model to the robot.

We also attach poses to locations and add a motion requirement to the `move` operator. The idea is that to move from `l_from` to `l_to` the robot will have to follow a path in `map1`.

```python
map1 = problem.add_map("./maps/my-map.yaml")
frame1 = problem.add_robot_frame([(-1.0, 0.5), (1.0, 0.5), (1.0, -0.5), (-1.0, -0.5)])
problem.attach_frame(r1, frame1)
problem.attach_motion_model(r1, ReedsSheppCar(turning_radius=4.0))
problem.attach_pose(l1, (4.0, 4.0, 0.0))
problem.attach_pose(l2, (26.0, 3.6, 0.0))
move.add_motion_requirement(robot, map1, l_from, l_to)
```

The frame may be moved into the motion model for simplicity (this would require changing the simple class above):

```python
map1 = problem.add_map("./maps/my-map.yaml")
problem.attach_motion_model(r1, ReedsSheppCar(frame=[(-1.0, 0.5), (1.0, 0.5), (1.0, -0.5), (-1.0, -0.5)],  turning_radius=4.0))
problem.attach_pose(l1, (4.0, 4.0, 0.0))
problem.attach_pose(l2, (26.0, 3.6, 0.0))
move.add_motion_requirement(robot, map1, l_from, l_to)
```

Solutions should also reflect the outcomes of motion planning. For this we could attach a look-up from action instances to motion requirements and paths. (There may be more than one motion requirement for a single action.)

**Comment by mikand** [comment]
@uwe-koeckemann Thanks for this proposal! We discussed this with @ElisaTosello and we came up with the following idea that seems somewhat cleaner to us. Let us know what you think! 

The new proposal should (:smile:) have the following advantages:
- The TAMP-related classes are in a dedicated package and subclass of `Problem`
- It is impossible to forget to attach a pose to a symbolic location or to not add a motion model, because we can check all these properties at construction-time
- It should be relatively easy to add the new features on top of the existing UP
- Does not use custom classes such as `ReedsSheppCar`, but an enumeration, because we need to ensure that the engine supports these features (we shall add this information to the `ProblemKind`)

```python
from unified_planning.model import Fluent, InstantaneousMotionAction, Object

#Note that we moved all TAMP-related things into a separate package
from unified_planning.model.tamp import TAMPProblem, MapType, PoseObject, MoveableType, MoveableObject, MotionModels
from unified_planning.shortcuts import UserType, BoolType, ObjectExp

# PoseType inherits from UserType and allows to discriminate plain objects from objects attachable to poses in a map
Location = MapType("location", "./maps/my-map.yaml")

# MoveableType inherits from UserType and imposes that all objects of this type are MoveableObject instances
Robot = MoveableType("robot")
robot_at = Fluent("robot_at", BoolType(), robot=Robot, position=Location)

l1 = PoseObject("l1", Location, (4.0, 4.0, 0.0))
l2 = PoseObject("l2", Location, (26.0, 3.6, 0.0))

r1 = MoveableObject("r1", Robot,
                    motion_model=MotionModels.REEDSSHEPP,
                    parameters={'frame': [(-1.0, 0.5), (1.0, 0.5), (1.0, -0.5), (-1.0, -0.5)],  'turning_radius': 4.0})

move = InstantaneousMotionAction("move", robot=Robot, l_from=Location, l_to=Location)
robot = move.parameter("robot")
l_from = move.parameter("l_from")
l_to = move.parameter("l_to")
move.add_precondition(robot_at(robot, l_from))
move.add_effect(robot_at(robot, l_from), False)
move.add_effect(robot_at(robot, l_to), True)

# NOTE: add_motion_requirement(moveable: FNode, map: TAMPMap, current_pose: FNode, waypoints: List[FNode])
# checks that:
# - the type of `moveable` is a TAMPMoveableType
# - `current_pose` (the initial position of moveable) has type TAMPPoseType
# - every element of `waypoints` has type TAMPPoseType
# - `current_pose` and every element of `waypoints` have the same type (i.e. they belong to the same map)
move.add_motion_requirement(robot, l_from, [l_to])

problem = TAMPProblem("robot")
problem.add_fluent(robot_at)
problem.add_action(move)

# Here we must check that l1 and l2 are PoseObjects and not just objects because their type is MapType
problem.add_object(l1)
problem.add_object(l2)

# Here we must check that r1 is a MoveableObject and not just an object because its type is a MoveableType
problem.add_object(r1)
problem.set_initial_value(robot_at(l1), True)
problem.set_initial_value(robot_at(l2), False)
problem.add_goal(robot_at(l2))
```

@uwe-koeckemann What do you think? @felixfi Can you also have a look?

**Reply by uwe-koeckemann** [reply]
This seems reasonable.

With the `InstantaneousMotionAction` class I guess we will have to create another class if we wanted to allow a `DurativeMotionAction`?

**Reply by mikand** [reply]
Yep

**Comment by ElisaTosello** [comment]
@uwe-koeckemann Here is our final proposal wrt to your case study.  
Briefly:
- a movable object has a shape (model) of any type and may have a motion model;
- a movable object acts in a certain map and, while acting within it, it can take certain configurations (instead of simple poses), where a configuration includes the state of any DOF of the object itself;
- actions may have a motion constraint which is strictly related to the motion model of the moveable object involved in that action.

Please, see the comments' snippet for a better explanation. 

Let us know if it sounds reasonable for you!
We will take care of the implementation! (@alvalentini :) )

```
from unified_planning.model import Fluent, InstantaneousMotionAction, Object
from unified_planning.shortcuts import BoolType
​
#Note that we moved all TAMP-related things into a separate package
from unified_planning.model.tamp import MoveableType, ConfigurationType, OccupancyMap, ConfigurationObject, MoveableObject, MotionModels, TAMPProblem
​
# assumptions:
# 1. the world is deterministic
# 2. the world is completely known
# 3. the moveable object (e.g., the robot) moves in a map composed of fixed obstacles
# 4. there is one common reference system (e.g., /world (0.0, 0.0, 0.0)) - that is the reference system of the map
​
Robot = MoveableType("robot")
​
# representation of the state of a movable object
# the input is equals to the number of variables useful to define this state
# (e.g., 3 = [x, y, yaw] - N = [N-DOFs of a robot])
RobotConfig = ConfigurationType("robot_config", 5)
​
robot_at = Fluent("robot_at", BoolType(), robot=Robot, configuration=RobotConfig)
​
# representation of the free and occupied working space, fixed obstacles are located on the occupied areas (e.g., Octomap)
map = OccupancyMap("./maps/my-map.yaml")
​
# configurations in the map 
# map and RobotConfig added for consistency check
# e.g., `c1` is a configuration expressed via 5 variables and is a collision free configuration in `map`
c1 = ConfigurationObject("c1", map, RobotConfig, (4.0, 4.0, 0.0, 2, 3))
c2 = ConfigurationObject("c2", map, RobotConfig, (4.0, 4.0, 0.0, 2, 3))
​
r1 = MoveableObject("r1", Robot,
                    model="robot.urdf",
                    motion_model=MotionModels.REEDSSHEPP,
                    parameters={'turning_radius': 4.0})
​
move = InstantaneousMotionAction("move", robot=Robot, c_from=RobotConfig, c_to=RobotConfig)
robot = move.parameter("robot")
c_from = move.parameter("c_from")
c_to = move.parameter("c_to")
move.add_precondition(robot_at(robot, c_from))
move.add_effect(robot_at(robot, c_from), False)
move.add_effect(robot_at(robot, c_to), True)
​
# there exists a motion control in your motion model that lets the moveable object moves from c_from to [c_to], 
# where [c_to] is a set of waypoints in your map
move.add_motion_constraint(Motion(robot, c_from, [c_to]))
​
problem = TAMPProblem("robot")
problem.add_fluent(robot_at)
problem.add_action(move)
problem.add_object(c1)
problem.add_object(c2)
problem.add_object(r1)
problem.set_initial_value(robot_at(c1), True)
problem.set_initial_value(robot_at(c2), False)
problem.add_goal(robot_at(c2))
```

**Reply by uwe-koeckemann** [reply]
Thank you, this looks very good! I will fix things on my end with this example in mind.

**Reply by uwe-koeckemann** [reply]
Should the `MovableObject` also include a reference to the map?

**Reply by uwe-koeckemann** [reply]
I am trying to convert this to my representation by hand and maybe `map` should be part of `RobotConfig` instead which is part of `robot_at` and the `ConfigurationObject`?

Essentially when converting `move.add_motion_constraint(Motion(robot, c_from, [c_to]))`, I need the map in which this motion constraint is supposed to hold, but the parameters are of type `RobotConfig` and only concrete objects contain the `map` reference. It may also avoid strange situations where `c1`and `c2` refer to different maps.

So, is there any situation in which instances of the same `ConfigurationType` should refer to different maps?

**Reply by ElisaTosello** [reply]
Your proposal is absolutely reasonable!

I would make these changes:

- the map is part of ```RobotConfig```, so that you can access the map through ```RobotConfig```:
```
RobotConfig = ConfigurationType("robot_config", map, 5)
```
- the map has a reference frame:
```
map = OccupancyMap("./maps/my-map.yaml", reference_frame=())
```
In this way, we made a strong assumption: we have only one reference system - the map one - and movable object configurations are expressed via this reference system. If you need a ref for each movable object (separated from the map itself), we should add another object, e.g., a ```Scene``` which will collect all transformations between the movable objects and the map. Let me know!

Focusing on multiple maps, generally - maybe in the future - we would like to have the possibility of navigating between multiple connected maps (e.g., the maps of two rooms connected by a door), but it sounds reasonable that waypoints in these maps should be uniquely identifiable.

The rest of the snippet remains unchanged.
